### PR TITLE
[[ Bug 18309 ]] Implement filter where expression

### PR DESCRIPTION
**Pull request by montegoulding**
This patch implements filter with expression using the global `each` in
the same way as the sort command.

### CODE REVIEW

**Comment by bwmilby** [comment]
Is there any particular rule on when a syntax should be split into multiple lines vice stated on a single line using square brackets/braces?
This will be a nice addition.

**Comment by montegoulding** [comment]
Basically if it's different enough and the current syntax is already very complicated it's simpler for everyone if it's split... otherwise it would have to be:

```
filter [{lines | items | keys | elements} of] <filterSource> { {with | without | [not] matching} [{wildcard pattern | regex pattern}] <filterPattern> | where <whereExpression> } [into <targetContainer>]
```

**Comment by bwmilby** [comment]
Understand.  Pretty much what I was thinking in the thread although I did move the word "pattern" outside of the brace to cut it down by a word.
`[{wildcard | regex} pattern]`

**Comment by montegoulding** [comment]
@bwmilby I think that is actually more correct for what the parser does

**Comment by montegoulding** [comment]
rebased & changed to 9.5

File: docs/dictionary/command/filter.lcdoc [hunk 1]
```diff
@@ -6,6 +6,8 @@ Syntax: filter [{lines | items | keys | elements} of] <filterSource> {with| with
 
 Syntax: filter [{lines | items | keys | elements} of] <filterSource> {with| without | [not] matching} regex pattern <regexPattern> [into <targetContainer>]
```
Comment (runrevmark) [inline docs/dictionary/command/filter.lcdoc]: `pattern` appears to be optional (looking at the parsing code

File: engine/src/exec-strings.cpp [hunk 2]
```diff
@@ -1993,26 +1993,15 @@ void MCStringsExecFilterRegex(MCExecContext& ctxt, MCStringRef p_source, MCStrin
...
 }
 
-void MCStringsExecFilterWildcardIntoIt(MCExecContext& ctxt, MCStringRef p_source, MCStringRef p_pattern, bool p_without, bool p_lines)
+void MCStringsExecFilterExpression(MCExecContext& ctxt, MCStringRef p_source, MCExpression* p_expression, bool p_without, bool p_lines, MCStringRef &r_result)
 {
-    MCAutoStringRef t_result;
-    MCStringsExecFilterWildcard(ctxt, p_source, p_pattern, p_without, p_lines, &t_result);
+    // Create the pattern matcher
+    MCPatternMatcher *matcher;
+    matcher = new (nothrow) MCExpressionMatcher(p_expression, p_source, ctxt . GetStringComparisonType());
```
Comment (runrevmark) [inline engine/src/exec-strings.cpp]: There doesn't seem to be any need for this to be heap-allocated...
```
MCPatternMatcher matcher(p_expression, ...)
```
Should have the same effect (just pass as `&matcher`)

Comment (runrevmark) [reply engine/src/exec-strings.cpp]: The same is true for the FilterRegex and FilterWildcard cases - probably worth adding the same change to them (but as a separate commit for clarity).

File: engine/src/patternmatcher.cpp [hunk 3]
```diff
@@ -109,7 +113,7 @@ bool MCRegexMatcher::compile(MCStringRef& r_error)
     return true;
 }
 
-bool MCRegexMatcher::match(MCRange p_range)
+bool MCRegexMatcher::match(MCExecContext ctxt, MCRange p_range)
```
Comment (runrevmark) [inline engine/src/patternmatcher.cpp]: MCExecContexts should always be passed by reference (i.e. MCExecContext&)

File: engine/src/patternmatcher.cpp [hunk 4]
```diff
@@ -282,7 +286,7 @@ static bool MCStringsWildcardMatchNative(const char *s, uindex_t s_length, const
     return p_index == p_length;
 }
 
-bool MCWildcardMatcher::match(MCRange p_source_range)
+bool MCWildcardMatcher::match(MCExecContext ctxt, MCRange p_source_range)
```
Comment (runrevmark) [inline engine/src/patternmatcher.cpp]: MCExecContext&

File: engine/src/patternmatcher.cpp [hunk 5]
```diff
@@ -313,3 +317,71 @@ bool MCWildcardMatcher::match(MCExecContext ctxt, MCNameRef p_key, bool p_match_
...
+{
+}
+
+bool MCExpressionMatcher::compile(MCStringRef& r_error)
+{
+    // expression patterns are not compiled
+    return true;
+}
+
+bool MCExpressionMatcher::match(MCExecContext ctxt, MCRange p_source_range)
```
Comment (runrevmark) [inline engine/src/patternmatcher.cpp]: MCExecContext&

**Comment by runrevmark** [comment]
Okay so before adding on ExpressionMatcher its probably worth cleaning up / sorting out the existing matchers. (i.e. sorting out the MCExecContext& and the unnecessary heap allocation)... Then add the new matcher on top.

There's a further refinement too (although not one for right now as it will require a fair bit of reworking - and there is probably better ways too) - the MCPatternMatcher base-class is a bit overkill really - as from what I can see the two main Filter functions which take a matcher only use the 'match()' function - this means the classes are really just a closure, e.g.:
```
class MCStringMatcher
{
public:
  virtual bool match(MCExecContext& ctxt, MCRange range) = 0;
};
```

**Comment by runrevmark** [comment]
@livecode-vulcan review ok c4478c8

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok c4478c80615aae60c4ea2e579f96999659eeea84

**Comment by livecode-vulcan** [comment]
😎 test success c4478c80615aae60c4ea2e579f96999659eeea84

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/504)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/496)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/496)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/496)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/496)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2018)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/246)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/245)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2050)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2040)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2038)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2073)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2010)

**Pull request merged**